### PR TITLE
History-style sharding for bitmap indices 

### PR DESCRIPTION
**Pull request by AskAlexSharov**
Method `writeBitmapSharded` using not optimal algo to split big bitmap to shards - probably need do kind of binary search. 
But it works fast enough, tomorrow will try to improve/simplify it. 

4Kb shard size - increase bucket size 1%.

### CODE REVIEW

File: eth/stagedsync/stage_log_index.go [hunk 1]
```diff
@@ -231,20 +231,17 @@ func unwindLogIndex(tx ethdb.Database, from, to uint64, quitCh <-chan struct{})
...
-		}
-		sz += sz1
+		sz += uint64(m.SerializedSizeInBytes())
 	}
 	const memoryNeedsForKey = 32 * 2 // each key stored in RAM: as string ang slice of bytes
 	return uint64(len(bitmaps)*memoryNeedsForKey)+sz > uint64(memLimit)
 }
 
 func flushBitmaps(c ethdb.Cursor, inMem map[string]*gocroaring.Bitmap) error {
+	defer func(t time.Time) { fmt.Printf("dbutils.go:258: %s\n", time.Since(t)) }(time.Now())
```
Comment (AlexeyAkhunov) [inline eth/stagedsync/stage_log_index.go:244]: Is this still needed?

File: ethdb/bitmapdb/dbutils.go [hunk 2]
```diff
@@ -3,121 +3,152 @@ package bitmapdb
...
-		return err
-	}
-	return nil
-}
 
-func hotShardOverflow(c ethdb.Cursor, initialKey []byte, hotShardN uint16, hotV []byte) (shardNForDelta uint16, err error) {
-	if hotShardN == 0 { // no cold shards, create new hot from delta - it will turn hot to cold automatically
-		return 1, nil
-	}
+	if len(currentLastV) < int(ShardLimit) { // merge delta to last shard
```
Comment (AlexeyAkhunov) [inline ethdb/bitmapdb/dbutils.go:46]: Can you simply do `delta = gocroaring.Or(delta, last)` and then `return writeBitmapSharded(c, key, delta)` without the rest of this function? As far as I can see, the `writeBitmapSharded` should take care of all of that

File: ethdb/bitmapdb/dbutils.go [hunk 3]
```diff
@@ -130,18 +161,17 @@ func TruncateRange(c ethdb.Cursor, key []byte, from, to uint64) error {
...
-		if uint64(bm.Maximum()) < from {
-			break
-		}
-		noReasonToCheckNextShard := uint64(bm.Minimum()) <= from && uint64(bm.Maximum()) >= to
+		noReasonToCheckNextShard := (uint64(bm.Minimum()) <= from && uint64(bm.Maximum()) >= to) || binary.BigEndian.Uint32(k[len(k)-4:]) == ^uint32(0)
 
-		updated++
 		bm.RemoveRange(from, to)
 		if bm.GetCardinality() == 0 { // don't store empty bitmaps
 			err = c.DeleteCurrent()
```
Comment (AlexeyAkhunov) [inline ethdb/bitmapdb/dbutils.go:168]: Is this a dangerous operation?

**Comment by AlexeyAkhunov** [comment]
Reviewed and left some comments. Good change. Need to do some testing. Also, we might need to convert into ETL (better in separate PR) to see if it helps with higher latency devices

**Comment by AlexeyAkhunov** [comment]
Did some testing, happy to merge once you've looked at the comments

File: ethdb/bitmapdb/dbutils.go (hunk 3 above)
Comment (AskAlexSharov) [reply ethdb/bitmapdb/dbutils.go:168]: I better will use dedicated cursor here and later switch to ETL. 
But if I understand assertion problem correctly - it happens when you iterate over dupsort bucket and delete last duplicated value (or do something else what causing conversion page from leaf to non-leaf). But I hadn't reproduce it. Maybe need to do something like fuzz testing to reproduce it (randomly add/delete/iterate - but use what we know about IH bucket - values have different length, custom comparator used, bucket dupsorted).

**Pull request merged**